### DISCUSSION TITLE
How to write custom DIRAC script to distribute with jobs

### DISCUSSION THREAD

**Discussion by IgorPelevanyuk** [DIRACGrid/DIRAC#7114] [Q&A] [answered]
Hello, 

In version 8 the idea behind development of scripts changed a bit. 
It looks like it is necessary to include all developed scripts in special `console_scripts` variable in `setup.cfg`.
Without that the script will throw an error "Something is very wrong" from https://github.com/DIRACGrid/DIRAC/blob/1cd1173303b177dbfe44ede43d18d25514713301/src/DIRAC/Core/Base/Script.py#L55C40-L55C63 

My question is the following. How to write a custom python script that use DIRAC API and that may be distributed with a job to be executed during userjob execution?

**Comment by chaen** [comment]
Hi Igor,

The scripts in `console_scripts` are the ones you get when installing DIRAC (`pip install DIRAC`). 
Any script can make use of `DIRAC` without being in this list, but it needs to be shipped, in your input sandbox for example. 
That hasn't changed between v7 and v8.

**Comment by IgorPelevanyuk** [comment]
Hi Chris! 

But when I do that and try to do inside job the following example:
```
from DIRAC.Interfaces.API.Dirac import Dirac
D = Dirac()
replicas = D.getReplicas(lfn)
```
I receive error:
```
  File "./jinr-get-file", line 77, in form_xrdcp_command
    replicas = D.getReplicas(lfn)
               ^^^^^^^^^^^^^^^^^^
  File ".../DIRAC_9Awsz9pilot/diracos/lib/python3.11/site-packages/DIRAC/Interfaces/API/Dirac.py", line 901, in getReplicas
    repsResult = dm.getReplicas(lfns, active=active, preferDisk=preferDisk, diskOnly=diskOnly)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../DIRAC/DataManagementSystem/Client/DataManager.py", line 1667, in getReplicas
    res = self.fileCatalog.getReplicas(lfnChunk, allStatus=allStatus)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../DIRAC/Resources/Catalog/FileCatalog.py", line 158, in __getattr__
    raise AttributeError

2023-07-21 10:11:34 UTC Framework/FileCatalog ERROR: FileCatalog._getEligibleCatalogs: Failed to get file catalog configuration. Path /Resources/FileCatalogs does not exist or it's not a section
2023-07-21 10:11:34 UTC Framework/FileCatalog ERROR: Failed to create catalog objects
AttributeError()
```

**Reply by IgorPelevanyuk** [reply]
And I should add that somehow the `dirac-dms-get-file` command works correctly.

**Reply by chaen** [reply, marked as answer]
you are probably missing the usual magic lines at the very top

```python
from DIRAC import initialize
initialize()
```

**Reply by IgorPelevanyuk** [reply]
And it works! 

Thank you, Chris! 

I wonder, should it be here or not: https://dirac.readthedocs.io/en/rel-v8r0/DeveloperGuide/AddingNewComponents/DevelopingCommands/index.html

**Reply by chaen** [reply]
This is not really the place because these are actual `DIRAC` scripts and the magic is done by the `@Script` decorator. But you are right, there should be a doc about that
